File: python/evm_state/cursor.py
"""Decode pinned Substreams cursors and bind durable progress to native rows.

Compatibility: bstream a513e03b7ada/cursor.go, opaque 0c01d37ea308/keys.go.
The upstream key and nonce below are PUBLIC obfuscation constants, not secrets
or an authentication mechanism. A decoded cursor alone proves no database state.
"""
import base64
import hashlib
import json
from pathlib import Path
import re

from nacl.exceptions import CryptoError
from nacl.secret import SecretBox

from .files import atomic_json
from .proof import VerificationError

PUBLIC_KEY = bytes.fromhex("7bfcacee257409099edd2bb6a442638b559a80bfbfc0b9acdea0d8344b10eb00")
PUBLIC_NONCE = bytes.fromhex("261554c45ab9b752abad4f19c242605702d55a0d91616a1b")


def decode(token):
    try:
        if (not isinstance(token, str) or not 1 <= len(token) <= 2048
                or not re.fullmatch(r"[A-Za-z0-9_-]+={0,2}", token)):
            raise ValueError("invalid cursor length")
        payload = SecretBox(PUBLIC_KEY).decrypt(base64.b64decode(token, altchars=b"-_", validate=True), PUBLIC_NONCE)
        parts = payload.decode("ascii").split(":")
        if not parts or (parts[0] in {"c1", "c2"} and len(parts) != 6) or (parts[0] == "c3" and len(parts) != 8):
            raise ValueError("invalid cursor segments")
        if parts[0] not in {"c1", "c2", "c3"} or parts[1] not in {"1", "16", "17"}:
            raise ValueError("unsupported cursor format or step")
        def ref(offset):
            number, block_hash = parts[offset:offset + 2]
            if not re.fullmatch(r"0|[1-9][0-9]*", number) or int(number) >= 2**64 or not re.fullmatch(r"[0-9a-f]{64}", block_hash):
                raise ValueError("invalid BSC block reference")
            return {"number": int(number), "hash": "0x" + block_hash}
        block = ref(2)
        head = block if parts[0] == "c1" else ref(4)
        lib = block if parts[0] == "c2" else ref(4 if parts[0] == "c1" else 6)
        if lib != block or head["number"] < block["number"] or (head["number"] == block["number"] and head != block):
            raise ValueError("cursor is not on a finalized block")
        return {"block": block, "head": head, "lib": lib, "step": int(parts[1])}
    except (ValueError, IndexError, CryptoError) as error:
        raise VerificationError("invalid or non-finalized native cursor") from error


def validate(client, run, token):
    position = decode(token)
    block = position["block"]
    identity = run["identity"]
    if block["number"] < identity["start_block"]:
        raise VerificationError("native cursor precedes its run's start")
    rows = list(client.rows("SELECT hash,accounts,schema_version,producer_version FROM state_blocks FINAL "
                           "WHERE number={number:UInt64}", {"number": block["number"]}))
    markers = list(client.rows("SELECT hash FROM _blocks_ FINAL WHERE number={number:UInt64}",
                              {"number": block["number"]}))
    if (len(rows) != 1 or rows[0]["hash"] != block["hash"] or rows[0]["accounts"] != ",".join(identity["accounts"])
            or int(rows[0]["schema_version"]) != 1 or int(rows[0]["producer_version"]) not in {3, 4, 5}
            or len(markers) != 1 or "0x" + markers[0]["hash"].removeprefix("0x") != block["hash"]):
        raise VerificationError("native cursor does not match its complete block data and marker")
    return position


def binding(run):
    identity_hash = hashlib.sha256(json.dumps(run["identity"], sort_keys=True, separators=(",", ":")).encode()).hexdigest()
    return {"format_version": 1, "run_id": run["run_id"], "database_uuid": run["database_uuid"],
            "identity_sha256": identity_hash}


def load_progress(client, run, directory):
    try:
        record = json.loads((Path(directory) / "durable_progress.json").read_text())
        if not isinstance(record, dict) or any(record.get(key) != value for key, value in binding(run).items()):
            raise VerificationError("durable progress belongs to another native run")
        position = validate(client, run, record["cursor"])
        if record.get("position") != position:
            raise VerificationError("durable progress position is corrupt")
        return record
    except (OSError, KeyError, TypeError) as error:
        raise VerificationError("durable progress is missing or invalid; restore matching run metadata") from error


def save_progress(client, run, directory, token):
    directory = Path(directory)
    position = validate(client, run, token)
    path = directory / "durable_progress.json"
    if path.exists():
        previous = load_progress(client, run, directory)
        if previous["position"]["block"]["number"] > position["block"]["number"]:
            raise VerificationError("native cursor regressed behind durable progress")
        if previous["cursor"] == token:
            return previous
    record = {**binding(run), "cursor": token, "position": position}
    atomic_json(path, record, overwrite=True)
    return record


def observe(client, run, directory):
    """Save a valid cursor while its native writer may replace the cursor file.

    A missing/torn native file is transient. Once decoded, mismatched database
    rows or a failed durable write are hard failures, not silently skipped.
    """
    try:
        with (Path(directory) / "cursor.txt").open() as handle:
            token = handle.read(2049).strip()
    except FileNotFoundError:
        return None
    try:
        decode(token)
    except VerificationError:
        return None
    return save_progress(client, run, directory, token)

File: crates/evm-state/src/cursor.rs
//! Decode the pinned bstream/opaque format. Public obfuscation, not authentication.
use crate::files::canonical_json;
use anyhow::{ensure, Context, Result};
use base64::{engine::general_purpose::URL_SAFE, Engine};
use crypto_secretbox::{
    aead::{Aead, KeyInit},
    XSalsa20Poly1305,
};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};

// PUBLIC upstream constants from opaque 0c01d37ea308; these are not secrets.
const PUBLIC_KEY: &str = "7bfcacee257409099edd2bb6a442638b559a80bfbfc0b9acdea0d8344b10eb00";
const PUBLIC_NONCE: &str = "261554c45ab9b752abad4f19c242605702d55a0d91616a1b";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct BlockRef {
    pub number: u64,
    pub hash: String,
}
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Position {
    pub block: BlockRef,
    pub head: BlockRef,
    pub lib: BlockRef,
    pub step: u8,
}

fn cipher() -> XSalsa20Poly1305 {
    XSalsa20Poly1305::new_from_slice(&hex::decode(PUBLIC_KEY).unwrap()).unwrap()
}

pub fn decode(token: &str) -> Result<Position> {
    ensure!(
        !token.is_empty() && token.len() <= 2048,
        "invalid cursor length"
    );
    let bare = token.trim_end_matches('=');
    ensure!(
        token.len() - bare.len() <= 2
            && !bare.is_empty()
            && bare
                .bytes()
                .all(|v| v.is_ascii_alphanumeric() || v == b'_' || v == b'-'),
        "invalid cursor alphabet"
    );
    let encoded = URL_SAFE.decode(token).context("invalid cursor base64")?;
    let nonce = hex::decode(PUBLIC_NONCE).unwrap();
    let payload = cipher()
        .decrypt(nonce.as_slice().into(), encoded.as_slice())
        .map_err(|_| anyhow::anyhow!("invalid cursor ciphertext"))?;
    let payload = std::str::from_utf8(&payload).context("invalid cursor text")?;
    let parts: Vec<_> = payload.split(':').collect();
    ensure!(
        matches!(parts.first(), Some(&"c1" | &"c2")) && parts.len() == 6
            || parts.first() == Some(&"c3") && parts.len() == 8,
        "invalid cursor segments"
    );
    ensure!(
        matches!(parts[1], "1" | "16" | "17"),
        "unsupported cursor step"
    );
    let reference = |offset: usize| -> Result<BlockRef> {
        let number = parts[offset];
        let hash = parts[offset + 1];
        ensure!(
            number == "0"
                || (!number.is_empty()
                    && !number.starts_with('0')
                    && number.bytes().all(|v| v.is_ascii_digit())),
            "invalid BSC block number"
        );
        ensure!(
            hash.len() == 64
                && hash
                    .bytes()
                    .all(|v| v.is_ascii_digit() || (b'a'..=b'f').contains(&v)),
            "invalid BSC block hash"
        );
        Ok(BlockRef {
            number: number.parse().context("block number exceeds uint64")?,
            hash: format!("0x{hash}"),
        })
    };
    let block = reference(2)?;
    let head = if parts[0] == "c1" {
        block.clone()
    } else {
        reference(4)?
    };
    let lib = if parts[0] == "c2" {
        block.clone()
    } else {
        reference(if parts[0] == "c1" { 4 } else { 6 })?
    };
    ensure!(
        lib == block
            && head.number >= block.number
            && (head.number != block.number || head == block),
        "cursor is not on a finalized block"
    );
    Ok(Position {
        block,
        head,
        lib,
        step: parts[1].parse()?,
    })
}

/// Compatibility fixture generation; the public cipher authenticates no user.
pub fn encode_public(payload: &str) -> Result<String> {
    let nonce = hex::decode(PUBLIC_NONCE).unwrap();
    let value = cipher()
        .encrypt(nonce.as_slice().into(), payload.as_bytes())
        .map_err(|_| anyhow::anyhow!("cursor encoding failed"))?;
    Ok(URL_SAFE.encode(value))
}

pub fn binding(run: &Value) -> Result<Value> {
    let identity = run.get("identity").context("missing native run identity")?;
    Ok(
        json!({"format_version":1,"run_id":crate::proof::string(run,"run_id")?,
        "database_uuid":crate::proof::string(run,"database_uuid")?,
        "identity_sha256":hex::encode(Sha256::digest(canonical_json(identity)?.as_bytes()))}),
    )
}

pub fn validate(client: &crate::ch::ClickHouse, run: &Value, token: &str) -> Result<Position> {
    use crate::ch::{params, uint};
    let position = decode(token)?;
    let identity = &run["identity"];
    ensure!(
        position.block.number >= uint(&identity["start_block"])?,
        "native cursor precedes its run's start"
    );
    let query = params(json!({"number":position.block.number}))?;
    let rows = client.rows("SELECT hash,accounts,schema_version,producer_version FROM state_blocks FINAL WHERE number={number:UInt64}", &query)?.collect::<Result<Vec<_>>>()?;
    let markers = client
        .rows(
            "SELECT hash FROM _blocks_ FINAL WHERE number={number:UInt64}",
            &query,
        )?
        .collect::<Result<Vec<_>>>()?;
    let accounts = identity["accounts"]
        .as_array()
        .context("missing native account filter")?
        .iter()
        .map(|v| v.as_str().context("invalid native account filter"))
        .collect::<Result<Vec<_>>>()?
        .join(",");
    ensure!(
        rows.len() == 1
            && rows[0]["hash"] == position.block.hash
            && rows[0]["accounts"] == accounts
            && uint(&rows[0]["schema_version"])? == 1
            && matches!(uint(&rows[0]["producer_version"])?, 3..=5)
            && markers.len() == 1
            && format!(
                "0x{}",
                crate::proof::string(&markers[0], "hash")?.trim_start_matches("0x")
            ) == position.block.hash,
        "native cursor does not match its complete block data and marker"
    );
    Ok(position)
}

pub fn load_progress(
    client: &crate::ch::ClickHouse,
    run: &Value,
    directory: &std::path::Path,
) -> Result<Value> {
    let record: Value = serde_json::from_slice(
        &std::fs::read(directory.join("durable_progress.json"))
            .context("durable progress is missing; restore matching run metadata")?,
    )?;
    let expected = binding(run)?;
    ensure!(
        expected
            .as_object()
            .unwrap()
            .iter()
            .all(|(key, value)| record.get(key) == Some(value)),
        "durable progress belongs to another native run"
    );
    let position = validate(client, run, crate::proof::string(&record, "cursor")?)?;
    ensure!(
        record["position"] == serde_json::to_value(position)?,
        "durable progress position is corrupt"
    );
    Ok(record)
}

pub fn save_progress(
    client: &crate::ch::ClickHouse,
    run: &Value,
    directory: &std::path::Path,
    token: &str,
) -> Result<Value> {
    let position = validate(client, run, token)?;
    let path = directory.join("durable_progress.json");
    if path.try_exists()? {
        let previous = load_progress(client, run, directory)?;
        ensure!(
            crate::ch::uint(&previous["position"]["block"]["number"])? <= position.block.number,
            "native cursor regressed behind durable progress"
        );
        if previous["cursor"] == token {
            return Ok(previous);
        }
    }
    let mut record = binding(run)?;
    record["cursor"] = json!(token);
    record["position"] = serde_json::to_value(position)?;
    crate::files::atomic_json(&path, &record, true)?;
    Ok(record)
}

pub fn observe(
    client: &crate::ch::ClickHouse,
    run: &Value,
    directory: &std::path::Path,
) -> Result<Option<Value>> {
    use std::io::Read;
    let file = match std::fs::File::open(directory.join("cursor.txt")) {
        Ok(file) => file,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(error) => return Err(error.into()),
    };
    let mut bytes = Vec::new();
    file.take(2049).read_to_end(&mut bytes)?;
    let Ok(token) = std::str::from_utf8(&bytes) else {
        return Ok(None);
    };
    let token = token.trim();
    if decode(token).is_err() {
        return Ok(None);
    }
    // After decoding, mismatched rows or a failed durable write are hard errors.
    save_progress(client, run, directory, token).map(Some)
}
